Add control flow nodes for implicit fields reads when reading a promoted field

owen-mc opened the pull request:
We need these to understand that if we write tainted data into a promoted field `outer.Inner.field` and read from `outer.field`, where `Inner` is an embedded field of `Outer`, then we get tainted data out.

This PR assumes that embedded fields are not pointer-typed. As follow-on work we should deal with this, either by introducing an analogue of `MkImplicitDeref` for `ImplicitFieldReadInstruction`s, or by making `ImplicitFieldReadInstruction` implicitly include a deref when needed.

smowton reviewed (commented):
Looks good -- tidying suggestions only; there might be a way to factor this more pleasantly but I don't see it right now, so happy in principle to merge once we've got good performance data.

sauyon commented:
I think there's an argument to be made that we shouldn't actually expose embedded fields the way we do, because technically according to the spec the field itself has the name of the embedded type, and the selector operator `.` does the chaining. But for backwards-compatibility's sake perhaps we should leave it as-is.

owen-mc commented:
@sauyon I don't understand your comment. Can you expand on "expose embedded fields the way we do"?

sauyon commented:
For a type

```
struct Foo {
  Bar
}
```

`foo.getField` will get the fields of `Bar` which isn't *technically* correct.

owen-mc commented on ql/src/semmle/go/controlflow/IR.qll:
```diff
@@ -253,21 +261,62 @@ module IR {
+      baseType =
+        se.getBase().getType().getUnderlyingType().(PointerType).getBaseType().getUnderlyingType()
+    |
+      field = baseType.getField(_) and
```
You mean this line is redundant because both the `if` and the `else` branches imply it? Yes, I think that's right. I've removed it now.

owen-mc commented:
I see what you mean. That would pretty much just amount to renaming `getOwnField` as `getField` and `getField` as `getFieldOrPromotedField`. To satisfy the principle of least surprise I think the current way we have it is better: we probably want people to use the one that can get promoted field when they don't know about the distinction.

owen-mc commented on ql/src/semmle/go/dataflow/SSA.qll:
```diff
@@ -319,9 +319,9 @@ private TSsaWithFields accessPath(IR::Instruction insn) {
-  exists(IR::FieldReadInstruction fr, IR::EvalInstruction frb |
+  exists(IR::FieldReadInstruction fr, IR::Instruction frb |
     fr.getBase() = frb or
-    fr.getBase() = IR::implicitDerefInstruction(frb.getExpr())
```
I'm not that familiar with this code, and I'm not sure I understand exactly what you're asking. The change to this method allows `frb` to be an `ImplicitFieldReadInstruction`, which `getBase()` can now return. It still deals with the two types that `getBase()` could already return: `EvalInstruction` or `EvalImplicitDerefInstruction` (which it looks through when defining `frb`).

More generally, this method is used to define `SsaWithFields`, and I think the situation is better now. Before there was an `SsaWithFields` for `(def@90:6).field` and also one for `(def@90:6).Middle.Inner.field`, even though those represent the same field, but now there is only `(def@90:6).Middle.Inner.field`.

owen-mc replied:
Now I think about it, maybe you were asking what `EvalImplicitDerefInstruction`s exist now. That hasn't changed at all. We might end up expanding that as part of follow-on work to deal with embedded fields with pointer type.

smowton commented:
At least we definitely should maintain compatibility for existing users

smowton commented on ql/src/semmle/go/controlflow/IR.qll:
```diff
@@ -253,21 +261,52 @@ module IR {
 
+  private Instruction selectorBase(SelectorExpr se, Field field) {
+    exists(StructType baseType |
+      baseType = se.getBase().getType().getUnderlyingType() or
```
Name `se.getBase().getType().getUnderlyingType()`

smowton commented on ql/src/semmle/go/controlflow/ControlFlowGraphImpl.qll:
```diff
@@ -1712,11 +1742,28 @@ module CFG {
         or
         not exists(MkImplicitDeref(this.getBase())) and
-        succ = mkExprOrSkipNode(this)
+        (
```
^^ still TODO

owen-mc replied:
Oops, I forgot I left this to do at the end. I like the approach you suggested, which I haven't seen before.

owen-mc commented:
I think I've addressed all the review comments. Now I'll look into why most (all?) of the tests are failing. After that I'll set off a dist-compare run to evaluate the effect on performance.

owen-mc commented:
I think I've found the cause of the test failures. When updating `selectorBase` to treat `SelectorExpr`s relating to field reads differently, I accidentally stopped it from dealing with `SelectorExpr`s which relate to method reads. (And it was only 50 tests that were failing.)

owen-mc commented:
The Revel tests were failing because they used reads of promoted fields that went through embedded fields with pointer types. I have added some commits to support this in the quickest way possible: `MkImplicitFieldSelection`s now also implicitly dereference the implicit field that's being selected if it has a pointer type. It would probably be more principled to either expand `MkImplicitDeref` to cover implicit derefs of this type of instruction or make a new control flow node for implicit dereferences of implicit field selections.

smowton reviewed (approved):
Looks good! Evaluation underway?

owen-mc commented:
I've just pushed lots of changes. Some are from https://github.com/owen-mc/codeql-go/pull/1, some are from https://github.com/github/codeql-go/pull/509, and some I've just added.

This needs a performance evaluation.

smowton commented:
@owen-mc test results look like we broke Revel somehow

smowton reviewed (approved):
This is looking great -- we might perhaps be able to unify the field and method paths a bit more, but if so that's strictly for a tidying followup. Especially like the new much-cleaner navigation of the stack of implicit reads in the IR layer.

Performance eval underway; intend to merge this if it's good.

smowton commented:
Eval results are good, will merge after a rebase

smowton commented:
```Project                                                c472bfcaa1a@87d8bc8d  c472bfcaa1a@e920f07a  time
go_reuseport                                           11                    10                    0.909091
lyft-go-samples                                        12                    11                    0.916667
uberalls                                               12                    11                    0.916667
assume-role-cli                                        14                    13                    0.928571
gostats                                                15                    14                    0.933333
carbonapi                                              30                    29                    0.966667
archiver                                               16                    16                    1
carbonzipper                                           15                    15                    1
cherami-client-go                                      16                    16                    1
go-torch                                               11                    11                    1
gojenkins                                              14                    14                    1
gonduit                                                12                    12                    1
goruntime                                              12                    12                    1
image_a3d69a4a89244803d2f5350aca6dd0fcbe444551         32                    32                    1
kiam                                                   23                    23                    1
oauth2_proxy_712739f7775378d39b8a6f6c8051188962062e9b  20                    20                    1
orangeforum_3f73f963cdf0b9093877711f0d84ff091683e452   18                    18                    1
pam-ussh                                               11                    11                    1
protoc-gen-star                                        17                    17                    1
smokescreen                                            11                    11                    1
sqlx                                                   13                    13                    1
tcheck                                                 12                    12                    1
thrift_264a3f318ed3e9e51573f67f963c8509786bcec2        22                    22                    1
thrift_6e5c0f6e315ea1cd8526789558bfd10d6cee2173        22                    22                    1
client-go_66e83da33c604751671a39f929644b8f471e5cd9     97                    98                    1.01031
rclone                                                 235                   238                   1.01277
aws-sdk-go                                             1453                  1487                  1.0234
terraform                                              163                   167                   1.02454
ngrok                                                  40                    41                    1.025
dns_501e858f679edecd4a38a86317ce50271014a80d           33                    34                    1.0303
hydra_9b5bbd48a72096930af08402c5e07fce7dd770f3         33                    34                    1.0303
makisu                                                 33                    34                    1.0303
cadvisor                                               32                    33                    1.03125
hydra_7f50b944ea7a0bc02f37a08c860670bd33453986         32                    33                    1.03125
image_8ad47b442feae0a21bff51a6058227fa1e45bb41         32                    33                    1.03125
client-go_689d090711762a4ab315320d706d3c9e3b225d3d     93                    96                    1.03226
prototool                                              31                    32                    1.03226
storagetapper                                          31                    32                    1.03226
vitess                                                 591                   611                   1.03384
kubernetes_7fbfe70147f963bc1fd2b0376d7efc116cc757a8    1562                  1617                  1.03521
tilt                                                   53                    55                    1.03774
authenticator                                          25                    26                    1.04
kubernetes                                             769                   801                   1.04161
ringpop-go                                             24                    25                    1.04167
cadence                                                623                   649                   1.04173
gobot_9a9d580de37056c636751930c9b33c03c294d3a9         47                    49                    1.04255
beego                                                  70                    73                    1.04286
cherami-server                                         70                    73                    1.04286
flytestdlib                                            23                    24                    1.04348
gobot_9b97c70ca6bed3ef922d8e43a01e4524e956a04a         46                    48                    1.04348
sourcegraph_bde46f4ae891ca5bccaab042cab8fa72f6c2b165   156                   163                   1.04487
go-blessclient                                         21                    22                    1.04762
aresdb                                                 101                   106                   1.0495
go-jose_d00415a0a4fdbcfbdf69deae1fe07fc953d9e76d       20                    21                    1.05
gorm                                                   20                    21                    1.05
tchannel-go                                            39                    41                    1.05128
nomad                                                  232                   244                   1.05172
go-jose_789a4c4bd4c118f7564954f441b29c153ccd6a96       19                    20                    1.05263
oauth2_proxy_289a6ccf463a425c7606178c510fc5eeb9c8b050  19                    20                    1.05263
zanzibar                                               147                   155                   1.05442
concourse_fabee1733ee9d084604c909fff2a015dcb21dd8e     128                   135                   1.05469
cni-ipvlan-vpc-k8s                                     18                    19                    1.05556
orangeforum_1f6313cb3a1e755880fc1354f3e1efc4dd2dd4aa   18                    19                    1.05556
kubernetes_47063891dd782835170f500a83f37cc98c3c1013    1530                  1619                  1.05817
cockroach                                              3310                  3504                  1.05861
plugins                                                17                    18                    1.05882
dns_833bf76c282d338e307ff7ec181b95cfc117deb2           33                    35                    1.06061
sourcegraph_2555ef0e46d4b0c457cb1894d63773af48821388   157                   167                   1.06369
astro                                                  15                    16                    1.06667
concourse_38cb4cc025e5ed28764b4adc363a0bbf41f3c7cb     131                   140                   1.0687
gogs_1f247cf8139cb483276cd8dd06385a800ce9d4b2          72                    77                    1.06944
gogs_c9bb33afc3ae35db21b26fd914bd80ca277a4e0d          71                    76                    1.07042
arachne                                                14                    15                    1.07143
concourse_dc3d15ab6c3a69890c9985f9c875d4c2949be727     126                   135                   1.07143
concourse_091671e19b3779e439f5ad4a6b4b89aa20a33778     131                   141                   1.07634
lyft-go-sdk                                            13                    14                    1.07692
source-to-image                                        26                    28                    1.07692
carbonserver                                           12                    13                    1.08333
spark-on-k8s-operator                                  23                    25                    1.08696
```
